modules/quanthub_core/src/AzureWorkloadIdentityTokenProvider.php

```diff
@@ -0,0 +1,231 @@
+  // Default cache filename.
+  private const DEFAULT_TOKEN_FILE = "default_workload_identity_token_file";
+  // Default Azure scope.
+  private const DEFAULT_SCOPE = 'https://ossrdbms-aad.database.windows.net/.default';
+  // Buffer time for token expiration (10 minutes)
+  private const EXPIRES_BEFORE_SEC = 600;
+  // Default Azure authority host.
+  private const DEFAULT_AUTHORITY_HOST = 'https://login.microsoftonline.com/';
+
+  /**
+   * Azure Client ID from environment or constructor.
+   *
+   * @var string */
+  private string $clientId;
+  /**
+   * Azure Tenant ID from environment or constructor.
+   *
+   * @var string */
+  private string $tenantId;
+  /**
+   * Path to federated token file from environment or constructor.
+   *
+   * @var string */
+  private string $federatedTokenFile;
+  /**
+   * Azure authority host URL.
+   *
+   * @var string */
+  private string $authorityHost;
+
+  /**
+   * Initialize token provider with configuration.
+   *
+   * @param string|null $clientId
+   *   Azure Client ID (default: AZURE_CLIENT_ID env)
+   * @param string|null $tenantId
+   *   Azure Tenant ID (default: AZURE_TENANT_ID env)
+   * @param string|null $federatedTokenFile
+   *   Path to federated token file (default: AZURE_FEDERATED_TOKEN_FILE env)
+   * @param string|null $authorityHost
+   *   Azure authority host URL (default: AZURE_AUTHORITY_HOST env
+   *   or DEFAULT_AUTHORITY_HOST)
+   */
+  public function __construct(
+    ?string $clientId = NULL,
+    ?string $tenantId = NULL,
+    ?string $federatedTokenFile = NULL,
+    ?string $authorityHost = NULL,
+  ) {
+    $this->clientId = $clientId ?? getenv('AZURE_CLIENT_ID') ?: '';
+    $this->tenantId = $tenantId ?? getenv('AZURE_TENANT_ID') ?: '';
+    $this->federatedTokenFile = $federatedTokenFile ?? getenv('AZURE_FEDERATED_TOKEN_FILE') ?: '';
+    $this->authorityHost = $authorityHost ?? getenv('AZURE_AUTHORITY_HOST') ?: self::DEFAULT_AUTHORITY_HOST;
+
+    $this->validateConfig();
+  }
+
+  /**
+   * Retrieve token data, using cached version if available and valid.
+   *
+   * @param string $scope
+   *   Azure scope for the token (default: DEFAULT_SCOPE)
+   * @param string|null $cachedTokenFileName
+   *   Custom cache filename (default: DEFAULT_TOKEN_FILE)
+   *
+   * @return array
+   *   Valid token data
+   *
+   * @throws RuntimeException
+   *   On configuration or token retrieval errors.
+   */
+  public function getToken(string $scope = self::DEFAULT_SCOPE, ?string $cachedTokenFileName = NULL): array {
+    $cacheFile = $cachedTokenFileName ?? self::DEFAULT_TOKEN_FILE;
+    return $this->getCachedToken($cacheFile) ?? $this->fetchNewToken($scope, $cacheFile);
+  }
+
+  /**
+   * Validate required configuration parameters.
+   *
+   * @throws RuntimeException
+   *   If any required configuration is missing.
+   */
+  private function validateConfig(): void {
+    if (empty($this->clientId) || empty($this->tenantId) || empty($this->federatedTokenFile)) {
```

Comment:
Please check 'USE_WORKLOAD_IDENTITY' too, looks like we would have this error on local machine.